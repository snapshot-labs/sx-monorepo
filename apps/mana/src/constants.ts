export const PORT = process.env.PORT || 3001;
