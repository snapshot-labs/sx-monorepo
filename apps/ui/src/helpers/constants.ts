import { VoteType, VoteTypeInfo } from '@/types';

export const APP_NAME = 'Snapshot';

export const SIDEKICK_URL = 'https://sh5.co';

export const ETH_CONTRACT = '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee';

export const CHAIN_IDS = {
  matic: 137,
  arb1: 42161,
  oeth: 10,
  eth: 1,
  sep: 11155111,
  'linea-testnet': 59140,
  bsc: 56,
  xdai: 100,
  fantom: 250,
  base: 8453
};

export const COINGECKO_ASSET_PLATFORMS = {
  1: 'ethereum',
  10: 'optimistic-ethereum',
  137: 'polygon-pos',
  42161: 'arbitrum-one'
};

export const COINGECKO_BASE_ASSETS = {
  1: 'ethereum',
  10: 'ethereum',
  137: 'matic-network',
  42161: 'ethereum'
};

export const MAX_SYMBOL_LENGTH = 12;

export const BASIC_CHOICES = ['For', 'Against', 'Abstain'];
export const SUPPORTED_VOTING_TYPES: VoteType[] = [
  'basic',
  'single-choice',
  'approval',
  'ranked-choice',
  'weighted',
  'quadratic'
] as const;
export const VOTING_TYPES_INFO: Record<
  Exclude<VoteType, 'custom'>,
  VoteTypeInfo
> = {
  basic: {
    label: 'Basic voting',
    description:
      'Voters have three choices: they can vote "For", "Against" or "Abstain".'
  },
  'single-choice': {
    label: 'Single choice voting',
    description: 'Voters can select only one choice from a predefined list.'
  },
  approval: {
    label: 'Approval voting',
    description:
      'Voters can select multiple choices, each choice receiving full voting power.'
  },
  'ranked-choice': {
    label: 'Ranked choice voting',
    description:
      'Each voter may select and rank any number of choices. Results are calculated by instant-runoff counting method.'
  },
  weighted: {
    label: 'Weighted voting',
    description:
      'Each voter may spread voting power across any number of choices.'
  },
  quadratic: {
    label: 'Quadratic voting',
    description:
      'Each voter may spread voting power across any number of choices. Results are calculated quadratically.'
  }
};
