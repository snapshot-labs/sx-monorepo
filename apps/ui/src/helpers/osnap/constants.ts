// contract addresses pulled from https://github.com/UMAprotocol/protocol/tree/master/packages/core/networks
export const contractData = [
  {
    // mainnet
    network: '1',
    name: 'OptimisticOracleV3',
    address: '0xfb55F43fB9F48F63f9269DB7Dde3BbBe1ebDC0dE',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/Bm3ytsa1YvcyFJahdfQQgscFQVCcMvoXujzkd3Cz6aof',
    deployBlock: 16636058
  },
  {
    // goerli
    network: '5',
    name: 'OptimisticOracleV3',
    address: '0x9923D42eF695B5dd9911D05Ac944d4cAca3c4EAB',
    subgraph:
      'https://api.thegraph.com/subgraphs/name/md0x/goerli-optimistic-oracle-v3',
    deployBlock: 8497481
  },
  {
    // optimism
    network: '10',
    name: 'OptimisticOracleV3',
    address: '0x072819Bb43B50E7A251c64411e7aA362ce82803B',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/FyJQyV5TLNeowZrL6kLUTB9JNPyWQNCNXJoxJWGEtBcn',
    deployBlock: 74537234
  },
  {
    // gnosis
    network: '100',
    name: 'OptimisticOracleV3',
    address: '0x22A9AaAC9c3184f68C7B7C95b1300C4B1D2fB95C',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/9K2nctaB2rAh7Cgzx3wKtdHwWoEeEQ9AThGATak6Ngm9',
    deployBlock: 27087150
  },
  {
    // polygon
    network: '137',
    name: 'OptimisticOracleV3',
    address: '0x5953f2538F613E05bAED8A5AeFa8e6622467AD3D',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/7KWbDhUE5Eqcfn3LXQtLbCfJLkNucnhzJLpi2jKhqNuf',
    deployBlock: 39331673
  },
  {
    //arbitrum
    network: '42161',
    name: 'OptimisticOracleV3',
    address: '0xa6147867264374F324524E30C02C331cF28aa879',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/9wpkM5tHgJBHYTzKEKk4tK8a7q6MimfS9QnW7Japa8hW',
    deployBlock: 61236565
  },
  {
    // avalanche
    network: '43114',
    name: 'OptimisticOracleV3',
    address: '0xa4199d73ae206d49c966cF16c58436851f87d47F',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/3k8gzGzTMV2vDZAGBFM2q642SUyVbE31bAUL8SjFQkre',
    deployBlock: 27816737
  },
  {
    // core
    network: '1116',
    name: 'OptimisticOracleV3',
    address: '0xD84ACa67d683aF7702705141b3C7E57e4e5e7726',
    subgraph:
      'https://thegraph.coredao.org/subgraphs/name/umaprotocol/core-optimistic-oracle-v3',
    deployBlock: 11341063
  },
  {
    // base
    network: '8453',
    name: 'OptimisticOracleV3',
    address: '0x2aBf1Bd76655de80eDB3086114315Eec75AF500c',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/2Q4i8YgVZd6bAmLyDxXgrKPL2B6QwySiEUqbTyQ4vm4C',
    deployBlock: 12066343
  },
  {
    // sepolia
    network: '11155111',
    name: 'OptimisticOracleV3',
    address: '0xFd9e2642a170aDD10F53Ee14a93FcF2F31924944',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/78JbrMhcC9CVDZHDADvNcyhRrrccTJG4vCVBztyer1Xa',
    deployBlock: 5421195
  },
  {
    // mainnet
    network: '1',
    name: 'OptimisticGovernor',
    address: '0x28CeBFE94a03DbCA9d17143e9d2Bd1155DC26D5d',
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/DQpwhiRSPQJEuc8y6ZBGsFfNpfwFQ8NjmjLmfv8kBkLu',
    deployBlock: 16890621
  },
  // Keep in mind, OG addresses are not the module addresses for each individual space, these addresses typically
  // are not used, but are here for reference.
  {
    //goerli
    network: '5',
    name: 'OptimisticGovernor',
    address: '0x07a7Be7AA4AaD42696A17e974486cb64A4daC47b',
    deployBlock: 8700589,
    subgraph:
      'https://api.thegraph.com/subgraphs/name/md0x/goerli-optimistic-governor'
  },
  {
    // optimism
    network: '10',
    name: 'OptimisticGovernor',
    address: '0x357fe84E438B3150d2F68AB9167bdb8f881f3b9A',
    deployBlock: 83168480,
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/Fd5RvSfkajAJ8Mi9sPxFSMVPFf56SDivDQW3ocqTAW5'
  },
  {
    // gnosis
    network: '100',
    name: 'OptimisticGovernor',
    deployBlock: 27102135,
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/RrkjZ6wTgLJkcjX68auzrEZHMRYwDx8kR5sFQQy4Phz'
  },
  {
    // polygon
    network: '137',
    name: 'OptimisticGovernor',
    address: '0x3Cc4b597E9c3f51288c6Cd0c087DC14c3FfdD966',
    deployBlock: 40677035,
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/7L2JM14PnZgxGnRX7xaz54zWS6KVK6ZqVRCxEKJrJTDG'
  },
  {
    // arbitrum
    network: '42161',
    name: 'OptimisticGovernor',
    address: '0x30679ca4ea452d3df8a6c255a806e08810321763',
    deployBlock: 72850751,
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/BfK867bnkQhnx1LspA99ypqiqxbAReQ92aZz66Ubv4tz'
  },
  {
    // avalanche
    network: '43114',
    name: 'OptimisticGovernor',
    address: '0xEF8b46765ae805537053C59f826C3aD61924Db45',
    deployBlock: 28050250,
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/5F8875fmvtnv8Vv4aeedUcwNWjuxUg54aTHdapFuMJi3'
  },
  {
    // core
    network: '1116',
    name: 'OptimisticGovernor',
    address: '0x596Fd6A5A185c67aBD1c845b39f593fBA9C233aa',
    deployBlock: 11341122,
    subgraph:
      'https://thegraph.coredao.org/subgraphs/name/umaprotocol/core-optimistic-governor'
  },
  {
    // base
    network: '8453',
    name: 'OptimisticGovernor',
    address: '0x80bCA2E1c272239AdFDCdc87779BC8Af6E12e633',
    deployBlock: 13062540,
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/H1WyWZqh5pHebWRDCXs7GhvGj7XznSP7arPY6pYcCqLn'
  },
  {
    // sepolia
    network: '11155111',
    name: 'OptimisticGovernor',
    address: '0x40153DdFAd90C49dbE3F5c9F96f2a5B25ec67461',
    deployBlock: 5421242,
    subgraph:
      'https://subgrapher.snapshot.org/subgraph/arbitrum/5pwrjCkpcpCd79k9MBS5yVgnsHQiw6afvXUfzqHjdRFw'
  }
] as const;

export const SAFE_APP_URLS = {
  '1': 'https://app.safe.global/apps/open',
  '5': 'https://app.safe.global/apps/open',
  '56': 'https://app.safe.global/apps/open',
  '100': 'https://app.safe.global/apps/open',
  '137': 'https://app.safe.global/apps/open',
  '246': 'https://app.safe.global/apps/open',
  '1116': 'https://safe.coredao.org/apps/open',
  '8453': 'https://app.safe.global/apps/open',
  '42161': 'https://app.safe.global/apps/open',
  '73799': 'https://app.safe.global/apps/open',
  '11155111': 'https://app.safe.global/apps/open'
} as const;

export const SAFE_PREFIXES = {
  1: 'eth',
  5: 'gor',
  10: 'oeth',
  100: 'gno',
  137: 'matic',
  1116: 'core',
  8453: 'base',
  42161: 'arb1',
  43114: 'avax',
  11155111: 'sep'
} as const;
