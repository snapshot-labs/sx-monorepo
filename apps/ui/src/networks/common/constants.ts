import { Connector } from '@/networks/types';

export const EVM_CONNECTORS: Connector[] = [
  'injected',
  'walletconnect',
  'walletlink',
  'gnosis'
];

export const STARKNET_CONNECTORS: Connector[] = ['argentx'];
export const EDITOR_APP_NAME = 'snapshot-v2';
