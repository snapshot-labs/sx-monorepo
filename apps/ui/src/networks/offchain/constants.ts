import { VoteType } from '@/types';
import { Connector } from '../types';

export const AUTHS = {};
export const PROPOSAL_VALIDATIONS = {
  any: 'Any',
  basic: 'Basic',
  'passport-gated': 'Passport gated',
  arbitrum: 'Arbitrum',
  'karma-eas-attestation': 'Karma EAS Attestation',
  'only-members': 'Only members'
};
export const STRATEGIES = {};
export const EXECUTORS = {};
export const CONNECTORS: Connector[] = ['injected', 'walletconnect'];
export const EDITOR_AUTHENTICATORS = [];
export const EDITOR_PROPOSAL_VALIDATIONS = [];
export const EDITOR_VOTING_STRATEGIES = [];
export const EDITOR_PROPOSAL_VALIDATION_VOTING_STRATEGIES = [];
export const EDITOR_EXECUTION_STRATEGIES = [];
export const EDITOR_SNAPSHOT_OFFSET = 4;
export const EDITOR_VOTING_TYPES: VoteType[] = [
  'basic',
  'single-choice',
  'approval',
  'ranked-choice',
  'weighted',
  'quadratic'
];

export const DEFAULT_VOTING_DELAY = 60 * 60 * 24 * 7;
