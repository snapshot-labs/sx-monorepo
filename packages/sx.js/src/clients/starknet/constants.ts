export const ESTIMATE_FEE_OVERHEAD_PERCENT = 50;
