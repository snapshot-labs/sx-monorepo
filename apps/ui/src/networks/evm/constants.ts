import { AbiCoder } from '@ethersproject/abi';
import { Web3Provider } from '@ethersproject/providers';
import { StandardMerkleTree } from '@openzeppelin/merkle-tree';
import { clients, evmNetworks } from '@snapshot-labs/sx';
import { HELPDESK_URL, MAX_SYMBOL_LENGTH } from '@/helpers/constants';
import { pinGraph } from '@/helpers/pin';
import { getUrl, shorten } from '@/helpers/utils';
import { NetworkID, StrategyParsedMetadata, VoteType } from '@/types';
import { StrategyConfig } from '../types';
import IHBeaker from '~icons/heroicons-outline/beaker';
import IHClock from '~icons/heroicons-outline/clock';
import IHCode from '~icons/heroicons-outline/code';
import IHCube from '~icons/heroicons-outline/cube';
import IHLightningBolt from '~icons/heroicons-outline/lightning-bolt';
import IHPencil from '~icons/heroicons-outline/pencil';
import IHUserCircle from '~icons/heroicons-outline/user-circle';

export function createConstants(networkId: NetworkID) {
  const config = evmNetworks[networkId];
  if (!config) throw new Error(`Unsupported network ${networkId}`);

  const SUPPORTED_AUTHENTICATORS = {
    [config.Authenticators.EthSig]: true,
    [config.Authenticators.EthTx]: true
  };

  const CONTRACT_SUPPORTED_AUTHENTICATORS = {
    [config.Authenticators.EthTx]: true
  };

  const SUPPORTED_STRATEGIES = {
    [config.Strategies.Vanilla]: true,
    [config.Strategies.Comp]: true,
    [config.Strategies.OZVotes]: true,
    [config.Strategies.Whitelist]: true
  };

  const SUPPORTED_EXECUTORS = {
    SimpleQuorumAvatar: true,
    SimpleQuorumTimelock: true,
    Axiom: true,
    Isokratia: true
  };

  const RELAYER_AUTHENTICATORS = {
    [config.Authenticators.EthSig]: 'evm'
  } as const;

  const AUTHS = {
    [config.Authenticators.EthSig]: 'Ethereum signature',
    [config.Authenticators.EthTx]: 'Ethereum transaction'
  };

  const PROPOSAL_VALIDATIONS = {
    [config.ProposalValidations.VotingPower]: 'Voting power'
  };

  const STRATEGIES = {
    [config.Strategies.Vanilla]: 'Vanilla',
    [config.Strategies.Comp]: 'ERC-20 Votes Comp (EIP-5805)',
    [config.Strategies.OZVotes]: 'ERC-20 Votes (EIP-5805)',
    [config.Strategies.Whitelist]: 'Merkle whitelist'
  };

  const EXECUTORS = {
    SimpleQuorumAvatar: 'Safe module (Zodiac)',
    SimpleQuorumTimelock: 'Timelock',
    Axiom: 'Axiom',
    Isokratia: 'Isokratia'
  };

  const EDITOR_AUTHENTICATORS = [
    {
      address: config.Authenticators.EthTx,
      name: 'Ethereum transaction',
      about:
        'Will authenticate a user by checking if the caller address corresponds to the author or voter address.',
      icon: IHCube,
      paramsDefinition: null
    },
    {
      address: config.Authenticators.EthSig,
      name: 'Ethereum signature',
      about:
        'Will authenticate a user based on an EIP-712 message signed by an Ethereum private key.',
      icon: IHPencil,
      paramsDefinition: null
    }
  ];

  const EDITOR_PROPOSAL_VALIDATIONS = [
    {
      address: config.ProposalValidations.VotingPower,
      type: 'VotingPower',
      name: 'Voting power',
      icon: IHLightningBolt,
      validate: (params: Record<string, any>) => {
        return params?.strategies?.length > 0;
      },
      generateSummary: (params: Record<string, any>) => `(${params.threshold})`,
      generateParams: (params: Record<string, any>) => {
        const abiCoder = new AbiCoder();

        const strategies = params.strategies.map((strategy: StrategyConfig) => {
          return {
            addr: strategy.address,
            params: strategy.generateParams
              ? strategy.generateParams(strategy.params)[0]
              : '0x00'
          };
        });

        return [
          abiCoder.encode(
            ['uint256', 'tuple(address addr, bytes params)[]'],
            [params.threshold, strategies]
          )
        ];
      },
      generateMetadata: async (params: Record<string, any>) => {
        const strategiesMetadata = await Promise.all(
          params.strategies.map(async (strategy: StrategyConfig) => {
            if (!strategy.generateMetadata) return;

            const metadata = await strategy.generateMetadata(strategy.params);
            const pinned = await pinGraph(metadata);

            return `ipfs://${pinned.cid}`;
          })
        );

        return {
          strategies_metadata: strategiesMetadata
        };
      },
      parseParams: async (params: string) => {
        const abiCoder = new AbiCoder();

        return {
          threshold: abiCoder
            .decode(
              ['uint256', 'tuple(address addr, bytes params)[]'],
              params
            )[0]
            .toString()
        };
      },
      paramsDefinition: {
        type: 'object',
        title: 'Params',
        additionalProperties: false,
        required: ['threshold'],
        properties: {
          threshold: {
            type: 'integer',
            title: 'Proposal threshold',
            examples: ['1']
          }
        }
      }
    }
  ];

  const EDITOR_VOTING_STRATEGIES = [
    {
      address: config.Strategies.Vanilla,
      name: 'Vanilla',
      about:
        'A strategy that gives one voting power to anyone. It should only be used for testing purposes and not in production.',
      link: `${HELPDESK_URL}/en/articles/9839150-vanilla-voting-strategy`,
      icon: IHBeaker,
      generateMetadata: async (params: Record<string, any>) => ({
        name: 'Vanilla',
        properties: {
          symbol: params.symbol,
          decimals: 0
        }
      }),
      parseParams: async (
        params: string,
        metadata: StrategyParsedMetadata | null
      ) => {
        if (!metadata) throw new Error('Missing metadata');

        return {
          symbol: metadata.symbol
        };
      },
      paramsDefinition: {
        type: 'object',
        title: 'Params',
        additionalProperties: false,
        required: [],
        properties: {
          symbol: {
            type: 'string',
            maxLength: MAX_SYMBOL_LENGTH,
            title: 'Symbol',
            examples: ['e.g. VP']
          }
        }
      }
    },
    {
      address: config.Strategies.Whitelist,
      type: 'MerkleWhitelist',
      name: 'Whitelist',
      about:
        'A strategy that defines a list of addresses each with designated voting power, using a Merkle tree for verification.',
      link: `${HELPDESK_URL}/en/articles/9839118-whitelist-voting-strategy`,
      generateSummary: (params: Record<string, any>) => {
        const length =
          params.whitelist.trim().length === 0
            ? 0
            : params.whitelist
                .split(/[\n,]/)
                .filter((s: string) => s.trim().length).length;

        return `(${length} ${length === 1 ? 'address' : 'addresses'})`;
      },
      generateParams: (params: Record<string, any>) => {
        const whitelist = params.whitelist
          .split(/[\n,]/)
          .filter((s: string) => s.trim().length)
          .map((item: string) => {
            const [address, votingPower] = item.split(':').map(s => s.trim());

            return [address, BigInt(votingPower)];
          });

        const tree = StandardMerkleTree.of(whitelist, ['address', 'uint96']);

        const abiCoder = new AbiCoder();
        return [abiCoder.encode(['bytes32'], [tree.root])];
      },
      generateMetadata: async (params: Record<string, any>) => {
        const tree = params.whitelist
          .split(/[\n,]/)
          .filter((s: string) => s.trim().length)
          .map((item: string) => {
            const [address, votingPower] = item.split(':').map(s => s.trim());

            return {
              address,
              votingPower: votingPower
            };
          });

        const pinned = await pinGraph({ tree });

        return {
          name: 'Whitelist',
          properties: {
            symbol: params.symbol,
            decimals: 0,
            payload: `ipfs://${pinned.cid}`
          }
        };
      },
      parseParams: async (
        params: string,
        metadata: StrategyParsedMetadata | null
      ) => {
        if (!metadata) throw new Error('Missing metadata');

        const getWhitelist = async (payload: string) => {
          const metadataUrl = getUrl(payload);

          if (!metadataUrl) return '';

          const res = await fetch(metadataUrl);
          const { tree } = await res.json();
          return tree
            .map((item: any) => `${item.address}:${item.votingPower}`)
            .join('\n');
        };

        return {
          symbol: metadata.symbol,
          whitelist: metadata.payload
            ? await getWhitelist(metadata.payload)
            : ''
        };
      },
      paramsDefinition: {
        type: 'object',
        title: 'Params',
        additionalProperties: false,
        required: [],
        properties: {
          whitelist: {
            type: 'string',
            format: 'addresses-with-voting-power',
            title: 'Whitelist',
            examples: ['0x556B14CbdA79A36dC33FcD461a04A5BCb5dC2A70:40']
          },
          symbol: {
            type: 'string',
            maxLength: 6,
            title: 'Symbol',
            examples: ['e.g. VP']
          }
        }
      }
    },
    {
      address: config.Strategies.OZVotes,
      name: 'ERC-20 Votes (EIP-5805)',
      about:
        'A strategy that allows delegated balances of OpenZeppelin style checkpoint tokens to be used as voting power.',
      link: `${HELPDESK_URL}/en/articles/9839125-erc-20-votes-eip-5805-voting-strategy`,
      icon: IHCode,
      generateSummary: (params: Record<string, any>) =>
        `(${shorten(params.contractAddress)}, ${params.decimals})`,
      generateParams: (params: Record<string, any>) => [params.contractAddress],
      generateMetadata: async (params: Record<string, any>) => ({
        name: 'ERC-20 Votes (EIP-5805)',
        properties: {
          symbol: params.symbol,
          decimals: parseInt(params.decimals),
          token: params.contractAddress
        }
      }),
      parseParams: async (
        params: string,
        metadata: StrategyParsedMetadata | null
      ) => {
        if (!metadata) throw new Error('Missing metadata');

        return {
          contractAddress: metadata.token,
          decimals: metadata.decimals,
          symbol: metadata.symbol
        };
      },
      paramsDefinition: {
        type: 'object',
        title: 'Params',
        additionalProperties: false,
        required: ['contractAddress', 'decimals'],
        properties: {
          contractAddress: {
            type: 'string',
            format: 'address',
            title: 'Token address',
            examples: ['0x0000…']
          },
          decimals: {
            type: 'integer',
            title: 'Decimals',
            examples: ['18']
          },
          symbol: {
            type: 'string',
            maxLength: MAX_SYMBOL_LENGTH,
            title: 'Symbol',
            examples: ['e.g. UNI']
          }
        }
      }
    },
    {
      address: config.Strategies.Comp,
      name: 'ERC-20 Votes Comp (EIP-5805)',
      about:
        'A strategy that allows delegated balances of Compound style checkpoint tokens to be used as voting power.',
      icon: IHCode,
      generateSummary: (params: Record<string, any>) =>
        `(${shorten(params.contractAddress)}, ${params.decimals})`,
      generateParams: (params: Record<string, any>) => [params.contractAddress],
      generateMetadata: async (params: Record<string, any>) => ({
        name: 'ERC-20 Votes Comp (EIP-5805)',
        properties: {
          symbol: params.symbol,
          decimals: parseInt(params.decimals),
          token: params.contractAddress
        }
      }),
      parseParams: async (
        params: string,
        metadata: StrategyParsedMetadata | null
      ) => {
        if (!metadata) throw new Error('Missing metadata');

        return {
          contractAddress: metadata.token,
          decimals: metadata.decimals,
          symbol: metadata.symbol
        };
      },
      paramsDefinition: {
        type: 'object',
        title: 'Params',
        additionalProperties: false,
        required: ['contractAddress', 'decimals'],
        properties: {
          contractAddress: {
            type: 'string',
            format: 'address',
            title: 'Token address',
            examples: ['0x0000…']
          },
          decimals: {
            type: 'integer',
            title: 'Decimals',
            examples: ['18']
          },
          symbol: {
            type: 'string',
            maxLength: MAX_SYMBOL_LENGTH,
            title: 'Symbol',
            examples: ['e.g. UNI']
          }
        }
      }
    }
  ];

  const EDITOR_PROPOSAL_VALIDATION_VOTING_STRATEGIES = EDITOR_VOTING_STRATEGIES;

  const EDITOR_EXECUTION_STRATEGIES = [
    {
      address: '',
      type: 'SimpleQuorumAvatar',
      name: EXECUTORS.SimpleQuorumAvatar,
      about:
        'An execution strategy that allows proposals to execute transactions from a specified target Avatar contract, the most popular one being a Safe.',
      icon: IHUserCircle,
      generateSummary: (params: Record<string, any>) =>
        `(${params.quorum}, ${shorten(params.contractAddress)})`,
      deploy: async (
        client: clients.EvmEthereumTx,
        web3: Web3Provider,
        controller: string,
        spaceAddress: string,
        params: Record<string, any>
      ): Promise<{ address: string; txId: string }> => {
        return client.deployAvatarExecution({
          signer: web3.getSigner(),
          params: {
            controller: params.controller,
            target: params.contractAddress,
            spaces: [spaceAddress],
            quorum: BigInt(params.quorum)
          }
        });
      },
      paramsDefinition: {
        type: 'object',
        title: 'Params',
        additionalProperties: false,
        required: ['controller', 'quorum', 'contractAddress'],
        properties: {
          controller: {
            type: 'string',
            format: 'address',
            title: 'Controller address',
            examples: ['0x0000…']
          },
          quorum: {
            type: 'integer',
            title: 'Quorum',
            examples: ['1']
          },
          contractAddress: {
            type: 'string',
            format: 'address',
            title: 'Avatar address',
            examples: ['0x0000…']
          }
        }
      }
    },
    {
      address: '',
      type: 'SimpleQuorumTimelock',
      name: EXECUTORS.SimpleQuorumTimelock,
      about:
        'Timelock implementation with a specified delay that queues proposal transactions for execution and includes an optional role to veto queued proposals.',
      icon: IHClock,
      generateSummary: (params: Record<string, any>) =>
        `(${params.quorum}, ${params.timelockDelay})`,
      deploy: async (
        client: clients.EvmEthereumTx,
        web3: Web3Provider,
        controller: string,
        spaceAddress: string,
        params: Record<string, any>
      ): Promise<{ address: string; txId: string }> => {
        return client.deployTimelockExecution({
          signer: web3.getSigner(),
          params: {
            controller: params.controller,
            vetoGuardian:
              params.vetoGuardian ||
              '0x0000000000000000000000000000000000000000',
            spaces: [spaceAddress],
            timelockDelay: BigInt(params.timelockDelay),
            quorum: BigInt(params.quorum)
          }
        });
      },
      paramsDefinition: {
        type: 'object',
        title: 'Params',
        additionalProperties: false,
        required: ['controller', 'quorum', 'timelockDelay'],
        properties: {
          controller: {
            type: 'string',
            format: 'address',
            title: 'Controller address',
            examples: ['0x0000…']
          },
          quorum: {
            type: 'integer',
            title: 'Quorum',
            examples: ['1']
          },
          vetoGuardian: {
            type: 'string',
            format: 'address',
            title: 'Veto guardian address',
            examples: ['0x0000…']
          },
          timelockDelay: {
            type: 'integer',
            format: 'duration',
            title: 'Timelock delay'
          }
        }
      }
    },
    {
      address: '',
      type: 'Axiom',
      name: EXECUTORS.Axiom,
      about:
        'This strategy enables offchain votes on the space. The validity of votes and voting power is verified onchain in bulk using a zkSNARK of storage proofs, which then triggers the execution of transactions.',
      icon: IHCode,
      generateSummary: (params: Record<string, any>) =>
        `(${shorten(params.contractAddress)}, ${params.slotIndex})`,
      deploy: async (
        client: clients.EvmEthereumTx,
        web3: Web3Provider,
        _controller: string,
        spaceAddress: string,
        params: Record<string, any>
      ): Promise<{ address: string; txId: string }> => {
        return client.deployAxiomExecution({
          signer: web3.getSigner(),
          params: {
            controller:
              params.controller || '0x0000000000000000000000000000000000000000',
            quorum: BigInt(params.quorum),
            contractAddress:
              params.contractAddress ||
              '0x0000000000000000000000000000000000000000',
            slotIndex: BigInt(params.slotIndex),
            space: spaceAddress,
            querySchema:
              '0xa09cc16ccaa32b96ca5c404c1b4be60d7883a7178f432e8f9f3c22157fc0f873'
          }
        });
      },
      paramsDefinition: {
        type: 'object',
        title: 'Params',
        additionalProperties: false,
        required: ['controller', 'quorum', 'contractAddress', 'slotIndex'],
        properties: {
          controller: {
            type: 'string',
            format: 'address',
            title: 'Controller address',
            examples: ['0x0000…']
          },
          quorum: {
            type: 'integer',
            title: 'Quorum',
            examples: ['1']
          },
          contractAddress: {
            type: 'string',
            format: 'address',
            title: 'Contract address',
            examples: ['0x0000…']
          },
          slotIndex: {
            type: 'integer',
            title: 'Slot index',
            examples: ['0']
          }
        }
      }
    },
    {
      address: '',
      type: 'Isokratia',
      name: EXECUTORS.Isokratia,
      about:
        'This strategy enables offchain votes on the space. The validity of votes is verified onchain in bulk using a zkSNARK, which then triggers the execution of transactions.',
      icon: IHCode,
      generateSummary: (params: Record<string, any>) =>
        `(${shorten(params.contractAddress)}, ${params.slotIndex})`,
      deploy: async (
        client: clients.EvmEthereumTx,
        web3: Web3Provider,
        _controller: string,
        _spaceAddress: string,
        params: Record<string, any>
      ): Promise<{ address: string; txId: string }> => {
        return client.deployIsokratiaExecution({
          signer: web3.getSigner(),
          params: {
            provingTimeAllowance: params.provingTimeAllowance,
            quorum: BigInt(params.quorum),
            queryAddress:
              params.queryAddress ||
              '0x0000000000000000000000000000000000000000',
            contractAddress:
              params.contractAddress ||
              '0x0000000000000000000000000000000000000000',
            slotIndex: BigInt(params.slotIndex)
          }
        });
      },
      paramsDefinition: {
        type: 'object',
        title: 'Params',
        additionalProperties: false,
        required: [
          'provingTimeAllowance',
          'quorum',
          'queryAddress',
          'contractAddress',
          'slotIndex'
        ],
        properties: {
          provingTimeAllowance: {
            type: 'integer',
            title: 'Proving time allowance',
            examples: ['3600']
          },
          quorum: {
            type: 'integer',
            title: 'Quorum',
            examples: ['1']
          },
          queryAddress: {
            type: 'string',
            format: 'address',
            title: 'Query address',
            examples: ['0x0000…']
          },
          contractAddress: {
            type: 'string',
            format: 'address',
            title: 'Contract address',
            examples: ['0x0000…']
          },
          slotIndex: {
            type: 'integer',
            title: 'Slot index',
            examples: ['0']
          }
        }
      }
    }
  ];

  const EDITOR_VOTING_TYPES: VoteType[] = ['basic'];

  return {
    SUPPORTED_AUTHENTICATORS,
    CONTRACT_SUPPORTED_AUTHENTICATORS,
    SUPPORTED_STRATEGIES,
    SUPPORTED_EXECUTORS,
    RELAYER_AUTHENTICATORS,
    AUTHS,
    PROPOSAL_VALIDATIONS,
    STRATEGIES,
    EXECUTORS,
    EDITOR_AUTHENTICATORS,
    EDITOR_PROPOSAL_VALIDATIONS,
    EDITOR_VOTING_STRATEGIES,
    EDITOR_PROPOSAL_VALIDATION_VOTING_STRATEGIES,
    EDITOR_EXECUTION_STRATEGIES,
    EDITOR_VOTING_TYPES
  };
}
