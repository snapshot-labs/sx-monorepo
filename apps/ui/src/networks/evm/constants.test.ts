import { describe, it, expect } from 'vitest';
import { createConstants } from './constants';

describe('EVM Constants', () => {
  const constants = createConstants('eth');
  const input = `0x395ed61716b48dc904140b515e9f682e33330154:1,0x395ed61716b48dc904140b515e9f682e33330154:2,0x395ed61716b48dc904140b515e9f682e33330154:3,
      0x395ed61716b48dc904140b515e9f682e33330154:4

  0x395ed61716b48dc904140b515e9f682e33330154:5,
,
  `;

  describe('EDITOR_VOTING_STRATEGIES', () => {
    describe('whitelist', () => {
      const strategy = constants.EDITOR_VOTING_STRATEGIES.find(s => s.name === 'Whitelist')!;

      describe('generateSummary', () => {
        it('should return the correct number of addresses', () => {
          const summary = strategy.generateSummary!({
            whitelist: input
          });

          expect(summary).toEqual('(5 addresses)');
        });
      });

      describe('generateParams', () => {
        it('should return the encoded value of the whitelist tree root', () => {
          const summary = strategy.generateParams!({
            whitelist: input
          });

          expect(summary).toMatchSnapshot();
        });
      });
    });
  });
});
