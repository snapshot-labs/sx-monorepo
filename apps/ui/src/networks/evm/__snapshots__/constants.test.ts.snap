// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EVM Constants > EDITOR_VOTING_STRATEGIES > whitelist > generateParams > should return the encoded value of the whitelist tree root 1`] = `
[
  "0x107aeb9f6cd74046422dc4138dd222efb6f0790d9e4e5e541fb00320e3866d49",
]
`;
